File: st-util/src/main/java/de/saring/util/unitcalc/SpeedToStringConverter.java
package de.saring.util.unitcalc;

import java.text.NumberFormat;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;


/**
 * Class for converting a speed (mostly the average speed) from the float value to a string value (both directions).
 * Depending on the specified SpeedMode the string value can be be either in km/h (or mph) or time/km or (time/mile),
 * the time is then in format mm:hh.
 * Unit system conversion (km or miles) will not be done here!
 *
 * @author Stefan Saring
 */
public class SpeedToStringConverter {

    private static final String ZERO_SPEED_TIME = "00:00";

    private SpeedMode speedMode;
    private final NumberFormat numberFormat;

    /**
     * Standard c'tor.
     *
     * @param speedMode speed mode to use
     */
    public SpeedToStringConverter(final SpeedMode speedMode) {

        if (speedMode != SpeedMode.SPEED && speedMode != SpeedMode.PACE) {
            throw new IllegalArgumentException("Invalid SpeedMode " + speedMode + "!");
        }

        this.speedMode = speedMode;
        this.numberFormat = NumberFormat.getInstance();
        this.numberFormat.setMaximumFractionDigits(3);
    }

    /**
     * Returns the currently used speed mode.
     *
     * @return speed mode
     */
    public SpeedMode getSpeedMode() {
        return speedMode;
    }

    /**
     * Sets the speed mode to be used for conversion.
     *
     * @param speedMode speed mode
     */
    public void setSpeedMode(SpeedMode speedMode) {
        this.speedMode = speedMode;
    }

    /**
     * Returns the specified speed value as a formatted string by using the defined speed mode.
     *
     * @param floatSpeed speed value as a float
     * @return the speed as string or null when conversion failed
     */
    public String floatSpeedtoString(final Float floatSpeed) {
        if (floatSpeed == null) {
            return null;
        }

        if (speedMode == SpeedMode.SPEED) {
            return numberFormat.format(floatSpeed);
        }
        else { // PACE
            if (floatSpeed == 0) {
                return ZERO_SPEED_TIME;
            }
            return FormatUtils.seconds2MinuteTimeString((int) (3600 / floatSpeed));
        }
    }

    /**
     * Returns the speed value as a float for the specified speed string by using the defined speed mode.
     *
     * @param strSpeed formatted speed string
     * @return the speed value or null when conversion failed
     */
    public Float stringSpeedToFloat(final String strSpeed) {
        try {
            final String strSpeedTrimmed = strSpeed.trim();

            if (speedMode == SpeedMode.SPEED) {
                return NumberFormat.getInstance().parse(strSpeed).floatValue();
            }
            else { // PACE
                if (ZERO_SPEED_TIME.equals(strSpeedTrimmed)) {
                    return 0f;
                }
                final LocalTime time = LocalTime.parse("00:" + strSpeedTrimmed, DateTimeFormatter.ISO_LOCAL_TIME);
                return 3600 / (float) (time.getMinute() * 60 + time.getSecond());
            }
        }
        catch (Exception e) {
            return null;
        }
    }
}

File: st-util/src/main/kotlin/de/saring/util/unitcalc/SpeedToStringConverter.kt
package de.saring.util.unitcalc

import java.text.NumberFormat
import java.time.LocalTime
import java.time.format.DateTimeFormatter


/**
 * Class for converting a speed (mostly the average speed) from the double value to a string value (both directions).
 * Depending on the specified SpeedMode the string value can be be either in km/h (or mph) or time/km or (time/mile),
 * the time is then in format mm:hh.
 * Unit system conversion (km or miles) will not be done here!
 *
 * @property speedMode speed mode to be used for the conversion
 *
 * @author Stefan Saring
 */
class SpeedToStringConverter(
        var speedMode: SpeedMode) {

    /**
     * Returns the specified speed value as a formatted string by using the defined speed mode.
     *
     * @param doubleSpeed speed value as a double
     * @return the speed as string or null when conversion failed
     */
    fun doubleSpeedtoString(doubleSpeed: Double?): String? {
        if (doubleSpeed == null) {
            return null
        }

        return if (speedMode === SpeedMode.SPEED) {
            NUMBER_FORMAT.format(doubleSpeed)
        } else { // PACE
            if (doubleSpeed == 0.0) {
                ZERO_SPEED_TIME
            } else {
                TimeUtils.seconds2MinuteTimeString((3600 / doubleSpeed).toInt())
            }
        }
    }

    /**
     * Returns the speed value as a double for the specified speed string by using the defined speed mode.
     *
     * @param strSpeed formatted speed string
     * @return the speed value or null when conversion failed
     */
    fun stringSpeedToDouble(strSpeed: String?): Double? {
        if (strSpeed == null) {
            return null
        }

        return try {
            val strSpeedTrimmed = strSpeed.trim()
            if (speedMode === SpeedMode.SPEED) {
                NumberFormat.getInstance().parse(strSpeed).toDouble()
            } else { // PACE
                val splittedPaceSpeed = strSpeedTrimmed.split(':')
                if (splittedPaceSpeed.size == 2) {
                    val minutes = splittedPaceSpeed[0].toInt()
                    val seconds = splittedPaceSpeed[1].toInt()
                    if (minutes == 0 && seconds == 0) {
                        0.0
                    } else {
                        3600 / (minutes * 60 + seconds).toDouble()
                    }
                } else {
                    null
                }
            }
        } catch (e: Exception) {
            null
        }
    }

    companion object {
        private const val ZERO_SPEED_TIME = "00:00"

        private val NUMBER_FORMAT = NumberFormat.getInstance().apply {
            maximumFractionDigits = 3
        }
    }
}
